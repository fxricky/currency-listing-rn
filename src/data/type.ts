export type Currency = {
  id: string;
  name: string;
  symbol: string;
  code?: string;
}
