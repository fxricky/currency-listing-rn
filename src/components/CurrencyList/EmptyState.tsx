import { icMagnifyGlass } from "@/assets/icons";
import colors from "@/themes/colors";
import React from "react";
import { StyleSheet, View, Text, Image } from "react-native";

type Props = {
  searchQuery: string;
};

export default function EmptyState({ searchQuery }: Props): React.ReactElement {
  return (
    <View style={styles.container} testID="currencylist#empty-state">
      <View style={styles.imgContainer}>
        <Image source={icMagnifyGlass} style={styles.imgMagnifyGlass} />
      </View>
      <Text style={styles.txtDescription}>
        {searchQuery
          ? `No results for "${searchQuery}". Try a different search term.`
          : "There are no currencies available in this list."}
      </Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    paddingHorizontal: 16,
  },
  imgContainer: {
    width: 60,
    height: 60,
    borderRadius: 50,
    backgroundColor: colors.gray["02"],
    alignItems: "center",
    justifyContent: "center",
  },
  imgMagnifyGlass: {
    width: 28,
    height: 28,
    tintColor: colors.gray["05"],
  },
  txtDescription: {
    fontSize: 16,
    color: colors.gray["06"],
    marginTop: 16,
    textAlign: "center",
  },
});
