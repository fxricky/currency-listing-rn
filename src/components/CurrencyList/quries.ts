export function isStartsWith(target: string, search: string) {
  return target.toLowerCase().startsWith(search.toLowerCase());
}

export function containLeadingWhitespace(target: string, search: string) {
  return target.includes(" " + search)
}
