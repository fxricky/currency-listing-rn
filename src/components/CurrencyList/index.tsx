import React, { createContext, useEffect, useMemo, useState } from "react";
import { FlatList, ListRenderItem, StyleSheet, View } from "react-native";
import CurrencyListItem from "@/components/CurrencyListItem";
import ListHeader from "@/components/ListHeader";
import { Currency } from "@/data/type";
import colors from "@/themes/colors";
import { isStartsWith, containLeadingWhitespace } from "./quries";
import EmptyState from "./EmptyState";

type Props = {
  displayList: Currency[];
  enableSearch?: boolean;
};

export const CurrencyListContext = createContext<{
  searchText: string;
  setSearchText: (v: string) => void;
}>({
  searchText: "",
  setSearchText: () => {},
});

function ItemSeparator(): React.ReactElement {
  return <View style={styles.separator} />;
}

export default function CurrencyList({
  displayList,
  enableSearch = false,
}: Props): React.ReactElement {
  // const [filteredList, setFilteredList] = useState<Currency[]>(displayList);
  const [searchText, setSearchText] = useState<string>("");

  const filteredList: Currency[] = useMemo(() => {
    const trimmedSearchText = searchText.trim();

    if (trimmedSearchText.length == 0) {
      return displayList;
    }

    return displayList.filter((item: Currency) => {
      return (
        isStartsWith(item.name, trimmedSearchText) ||
        isStartsWith(item.symbol, trimmedSearchText) ||
        containLeadingWhitespace(item.name, trimmedSearchText)
      );
    });
  }, [displayList, searchText]);

  const renderItem: ListRenderItem<Currency> = ({ item }) => {
    return <CurrencyListItem data={item} />;
  };

  return (
    <CurrencyListContext.Provider value={{ searchText, setSearchText }}>
      <ListHeader title="Currency List" enableSearch={enableSearch} />
      <FlatList
        data={filteredList}
        style={styles.container}
        renderItem={renderItem}
        ItemSeparatorComponent={() => <ItemSeparator />}
        ListEmptyComponent={() => <EmptyState searchQuery={searchText} />}
        contentContainerStyle={{
          flexGrow: 1,
        }}
      />
    </CurrencyListContext.Provider>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  separator: {
    height: 1,
    backgroundColor: colors.gray["02"],
  },
});
