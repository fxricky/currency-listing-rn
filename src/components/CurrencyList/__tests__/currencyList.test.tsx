import { render, screen } from "@testing-library/react-native";
import CurrencyList from "../index";
import { Currency } from "@/data/type";

const dummyData: Currency[] = [
  {
    id: "1",
    name: "Bitcoin",
    symbol: "BTC",
  },
  {
    id: "2",
    name: "Ethereum",
    symbol: "ETH",
  },
];

describe("CurrencyList", () => {
  test("renders correctly with default props", () => {
    const tree = render(<CurrencyList displayList={dummyData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("renders EmptyState when list is empty", () => {
    render(<CurrencyList displayList={[]} />);
    expect(screen.getByTestId("currencylist#empty-state")).toBeTruthy();
  });

  test("renders correctly with different data", () => {
    const differentData: Currency[] = [
      {
        id: "3",
        name: "Ripple",
        symbol: "XRP",
      },
    ];
    const tree = render(<CurrencyList displayList={differentData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("renders with search bar", () => {
    const tree = render(<CurrencyList displayList={dummyData} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
