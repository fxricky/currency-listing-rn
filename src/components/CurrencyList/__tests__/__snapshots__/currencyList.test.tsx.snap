// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyList renders correctly with default props 1`] = `
[
  <View
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#f3f4f6",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "maxWidth": "100%",
        "paddingHorizontal": 16,
        "paddingVertical": 16,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 20,
          "fontWeight": "bold",
          "lineHeight": 40,
        }
      }
    >
      Currency List
    </Text>
  </View>,
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    data={
      [
        {
          "id": "1",
          "name": "Bitcoin",
          "symbol": "BTC",
        },
        {
          "id": "2",
          "name": "Ethereum",
          "symbol": "ETH",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    style={
      {
        "flex": 1,
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "gap": 8,
              "justifyContent": "space-between",
              "paddingHorizontal": 16,
              "paddingVertical": 12,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#364153",
                "borderRadius": 16,
                "height": 32,
                "justifyContent": "center",
                "width": 32,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#f3f4f6",
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
            >
              B
            </Text>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <Text>
              Bitcoin
            </Text>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "gap": 8,
                }
              }
            >
              <Text>
                BTC
              </Text>
              <Image
                source={
                  {
                    "testUri": "../../../src/assets/icons/icChevronRight.png",
                  }
                }
                style={
                  {
                    "height": 24,
                    "tintColor": "#99a1af",
                    "width": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
        <View
          style={
            {
              "backgroundColor": "#f3f4f6",
              "height": 1,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "gap": 8,
              "justifyContent": "space-between",
              "paddingHorizontal": 16,
              "paddingVertical": 12,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#364153",
                "borderRadius": 16,
                "height": 32,
                "justifyContent": "center",
                "width": 32,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#f3f4f6",
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
            >
              E
            </Text>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <Text>
              Ethereum
            </Text>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "gap": 8,
                }
              }
            >
              <Text>
                ETH
              </Text>
              <Image
                source={
                  {
                    "testUri": "../../../src/assets/icons/icChevronRight.png",
                  }
                }
                style={
                  {
                    "height": 24,
                    "tintColor": "#99a1af",
                    "width": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;

exports[`CurrencyList renders correctly with different data 1`] = `
[
  <View
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#f3f4f6",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "maxWidth": "100%",
        "paddingHorizontal": 16,
        "paddingVertical": 16,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 20,
          "fontWeight": "bold",
          "lineHeight": 40,
        }
      }
    >
      Currency List
    </Text>
  </View>,
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    data={
      [
        {
          "id": "3",
          "name": "Ripple",
          "symbol": "XRP",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    style={
      {
        "flex": 1,
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "gap": 8,
              "justifyContent": "space-between",
              "paddingHorizontal": 16,
              "paddingVertical": 12,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#364153",
                "borderRadius": 16,
                "height": 32,
                "justifyContent": "center",
                "width": 32,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#f3f4f6",
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
            >
              R
            </Text>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <Text>
              Ripple
            </Text>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "gap": 8,
                }
              }
            >
              <Text>
                XRP
              </Text>
              <Image
                source={
                  {
                    "testUri": "../../../src/assets/icons/icChevronRight.png",
                  }
                }
                style={
                  {
                    "height": 24,
                    "tintColor": "#99a1af",
                    "width": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;

exports[`CurrencyList renders with search bar 1`] = `
[
  <View
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#f3f4f6",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "maxWidth": "100%",
        "paddingHorizontal": 16,
        "paddingVertical": 16,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 20,
          "fontWeight": "bold",
          "lineHeight": 40,
        }
      }
    >
      Currency List
    </Text>
  </View>,
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    data={
      [
        {
          "id": "1",
          "name": "Bitcoin",
          "symbol": "BTC",
        },
        {
          "id": "2",
          "name": "Ethereum",
          "symbol": "ETH",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    style={
      {
        "flex": 1,
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "gap": 8,
              "justifyContent": "space-between",
              "paddingHorizontal": 16,
              "paddingVertical": 12,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#364153",
                "borderRadius": 16,
                "height": 32,
                "justifyContent": "center",
                "width": 32,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#f3f4f6",
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
            >
              B
            </Text>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <Text>
              Bitcoin
            </Text>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "gap": 8,
                }
              }
            >
              <Text>
                BTC
              </Text>
              <Image
                source={
                  {
                    "testUri": "../../../src/assets/icons/icChevronRight.png",
                  }
                }
                style={
                  {
                    "height": 24,
                    "tintColor": "#99a1af",
                    "width": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
        <View
          style={
            {
              "backgroundColor": "#f3f4f6",
              "height": 1,
            }
          }
        />
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "gap": 8,
              "justifyContent": "space-between",
              "paddingHorizontal": 16,
              "paddingVertical": 12,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#364153",
                "borderRadius": 16,
                "height": 32,
                "justifyContent": "center",
                "width": 32,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#f3f4f6",
                  "fontSize": 16,
                  "fontWeight": "bold",
                }
              }
            >
              E
            </Text>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <Text>
              Ethereum
            </Text>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "gap": 8,
                }
              }
            >
              <Text>
                ETH
              </Text>
              <Image
                source={
                  {
                    "testUri": "../../../src/assets/icons/icChevronRight.png",
                  }
                }
                style={
                  {
                    "height": 24,
                    "tintColor": "#99a1af",
                    "width": 24,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;
